processor/src/tests/u32_ops.rs

```diff
@@ -0,0 +1,1009 @@
+
+    let inputs = build_inputs(inputs);
+    let trace = execute(&script, &inputs).unwrap();
+    let last_state = trace.last_stack_state();
+
+    let expected_state = build_stack_state(outputs);
+    assert_eq!(expected_state, last_state);
+}
+
+/// This helper function tests failures where the execution of a given assembly operation over the
+/// provided inputs is expected to panic. This function catches the panic and tests it against a
+/// provided string to make sure it contains the expected error string.
+fn test_execution_failure(asm_op: &str, inputs: &[u64], err_substr: &str) {
+    let script = compile(format!("begin {} end", asm_op).as_str());
+
+    let inputs = build_inputs(inputs);
+    assert_eq!(
+        std::panic::catch_unwind(|| execute(&script, &inputs).unwrap())
+            .err()
+            .and_then(|a| { a.downcast_ref::<String>().map(|s| s.contains(err_substr)) }),
+        Some(true)
+    );
+}
+
+/// This helper function tests failures where the compilation of a given assembly operation is
+/// expected to panic. This function catches the panic and tests it against a provided string to
+/// make sure it contains the expected error string.
+fn test_compilation_failure(asm_op: &str, err_substr: &str) {
+    assert_eq!(
+        std::panic::catch_unwind(|| compile(format!("begin {} end", asm_op).as_str()))
+            .err()
+            .and_then(|a| { a.downcast_ref::<String>().map(|s| s.contains(err_substr)) }),
+        Some(true)
+    );
+}
+
+/// This helper function tests a provided u32 assembly operation, which takes a single input, to
+/// ensure that it fails when the input is >= 2^32.
+fn test_input_out_of_bounds(asm_op: &str) {
+    test_execution_failure(asm_op, &[U32_BOUND], "FailedAssertion");
+}
+
+/// This helper function tests a provided u32 bitwise assembly operation which takes a parameter
+/// to ensure that it fails when that parameter is out of bounds above the maximum.
+fn test_param_out_of_bounds(asm_op_base: &str, over: u64) {
+    let build_asm_op = |param: u64| format!("{}.{}", asm_op_base, param);
+
+    // --- test b > max ---------------------------------------------------------------------------
+    test_compilation_failure(build_asm_op(over).as_str(), "parameter value");
+}
+
+/// This is a helper function which provides the expected result of a bitwise rotation operation
+/// for the specified rotation direction (left or right) and the given inputs a (the u32 to be
+/// rotated) and b (the amount of rotation).
+fn get_rotation_result(op_base: &str, a: u32, b: u32) -> u64 {
+    let rot_inverse = 32 - b;
+    match op_base {
+        "u32rotl" => {
+            // shift everything by b to get the upper bits of the result
+            let new_upper_bits = (a << b) as u32;
+            // move b bits which fall off the top during left shift to the result's low bits
+            let new_lower_bits = a >> rot_inverse;
+            (new_upper_bits | new_lower_bits) as u64
```

Comment:
Can't we use the built in [u32::rotate_left()](https://doc.rust-lang.org/std/primitive.u32.html#method.rotate_left) function here?